pkg/migration/migration.go

```diff
@@ -0,0 +1,573 @@
+package migration
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/libopenstorage/operator/drivers/storage"
+	corev1 "github.com/libopenstorage/operator/pkg/apis/core/v1"
+	"github.com/libopenstorage/operator/pkg/constants"
+	"github.com/libopenstorage/operator/pkg/controller/storagecluster"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	portworxDaemonSetName  = "portworx"
+	portworxContainerName  = "portworx"
+	migrationRetryInterval = 30 * time.Second
+	phaseAwaitingApproval  = "AwaitingMigrationApproval"
+)
+
+// Handler object that carries out migration of Portworx Daemonset
+// and it's components to operator managed StorageCluster object
+type Handler struct {
+	client client.Client
+	driver storage.Driver
+}
+
+// New creates a new instance of migration handler
+func New(ctrl *storagecluster.Controller) *Handler {
+	return &Handler{
+		client: ctrl.GetKubernetesClient(),
+		driver: ctrl.Driver,
+	}
+}
+
+// Start starts the migration
+func (h *Handler) Start() {
+	ds, err := h.getPortworxDaemonSet()
+	if err != nil {
+		logrus.Errorf("Migration failed to start. %v", err)
+		return
+	} else if ds == nil {
+		logrus.Infof("Migration is not needed")
+		return
+	}
+
+	cluster, err := h.createStorageClusterIfAbsent(ds)
+	if err != nil {
+		logrus.Errorf("Migration failed to create StorageCluster. %v", err)
+		return
+	}
+
+	wait.PollInfinite(migrationRetryInterval, func() (bool, error) {
+		cluster, isNeeded := h.isMigrationNeeded(cluster)
+		if !isNeeded {
+			return true, nil
+		} else if cluster == nil {
+			return false, nil
```

Comment:
In case of error we want to retry the migration as there could be something wrong with k8s api server. The `isMigrationNeeded` returns `true` and empty cluster on error instead of returning the error itself.
Returning the err from `isMigrationNeeded` was fine, but I get the same information from `cluster == nil`. Just wanted to avoid returning an extra argument from that function.

Reply:
ok literally when I see isMigrationNeeded and true I would think migration is needed, I think return false and and error is easier to understand. Also we dont need to check storagecluster here as it is already created above createStorageClusterIfAbsent right?

Reply:
I removed this and made it simple. If the daemonset is removed only then we stop the migration. If the storagecluster is deleted, we create a new one waiting for approval.